IEP-749 hints viewer based on the hints.yml

sigmaaa opened the pull request:
## Description

Fixes # ([IEP-749](https://jira.espressif.com:8443/browse/IEP-749))

## Type of change
- New feature (non-breaking change which adds functionality)

## How has this been tested?

Case 1:
- Go to Window -> Show View -> Other... -> Hints
- You will be able to see the Hints view like this:
![hints view](https://user-images.githubusercontent.com/24419842/185564327-dd686b9a-ed65-433a-a6c3-d052dd51ec49.png)
- In the search field you can type an error message, for example "../main/blink_example_main.c:16:10: fatal error: test.h: No such file or directory"
- after that, you will be able to see a filtered view with only one related hint
![filtered hints view](https://user-images.githubusercontent.com/24419842/185564953-2b9f346d-ca28-4da5-970e-f91b179d5d10.png)
- The hint may be too long to fit on the monitor. To quickly read the hint, double-click on the line to open the info dialog with full text
![hint details](https://user-images.githubusercontent.com/24419842/185565587-f904a184-e861-43e1-8ea6-3ec0cabdbeba.png)

**Test Configuration**:
* ESP-IDF Version: Master
* OS (Windows, Linux and macOS):

## Dependent components impacted by this PR:
- Hints view

## Checklist
- [x] PR Self Reviewed
- [x] Applied Code formatting
- [ ] Added Documentation
- [ ] Added Unit Test
- [ ] Verified on all platforms - Windows,Linux and macOS

AndriiFilippov commented:
hi @sigmaaa 

Tested under: 
OS - Windows 10 
ESP-IDF: master 

I don't see this filter works for me 

![image](https://user-images.githubusercontent.com/69584713/186356895-b3110951-00c6-464d-b4ef-156774260c50.png)

AndriiFilippov commented:
@sigmaaa 

Tested under:
OS - Linux
ESP-IDF: master

I was not even able to list hints.  (Window -> Show View -> Other... -> Hints)
Please, check. 

![image](https://user-images.githubusercontent.com/69584713/186365231-967179fc-b4b2-473d-b040-e1cf9f636340.png)

sigmaaa commented:
> hi @sigmaaa
> 
> Tested under: OS - Windows 10 ESP-IDF: master
> 
> I don't see this filter works for me
> 
> P.S. maybe we should change the hints logo also on Windows, to make users navigate better
> 
> ![image](https://user-images.githubusercontent.com/69584713/186357713-71d21bed-b28e-4cbd-856b-b7439314771a.png)

The filter works if an error message is complete. The idea is to find the error based on the regular expressions in the list.
Here is an example how real error message looks like: `../main/blink_example_main.c:16:10: fatal error: test.h: No such file or directory`. So the user just copy-paste this error message and will see a hint for it

AndriiFilippov commented:
> > hi @sigmaaa
> > Tested under: OS - Windows 10 ESP-IDF: master
> > I don't see this filter works for me
> > P.S. maybe we should change the hints logo also on Windows, to make users navigate better
> > ![image](https://user-images.githubusercontent.com/69584713/186357713-71d21bed-b28e-4cbd-856b-b7439314771a.png)
> 
> Hi @AndriiFilippov, thanks for the review. The filter works if an error message is complete. The idea is to find the error based on the regular expressions in the list. Here is an example how real error message looks like: `../main/blink_example_main.c:16:10: fatal error: test.h: No such file or directory`. So the user just copy-paste this error message and will see a hint for it
> 
> I will change the default "type filter text" message to the "type error message" to make it clear for the user

@sigmaaa ok, but I am not sure if it is a good idea to make filter works like this caz different users copy messages differently  and this is what exactly happened to me - I didn't know how does it work, so I used filter in my own way and because of this  I was sure - filter is broken.   
It would have more logic if we add some note for user - how to use this filter OR we should change the way filter works. 
@kolipakakondal  what do you think ?

sigmaaa commented:
> @sigmaaa
> 
> Tested under: OS - Linux ESP-IDF: master
> 
> I was not even able to list hints. (Window -> Show View -> Other... -> Hints) Please, check.
> 
> ![image](https://user-images.githubusercontent.com/69584713/186365231-967179fc-b4b2-473d-b040-e1cf9f636340.png)

Hi @AndriiFilippov pushed a new commit. Should work on Linux now. However, I noticed, that the hints.yml structure is updated on the latest esp-idf master. I need to make some changes to support the new hints.yml file. I'll keep you updated. Thanks

AndriiFilippov commented:
> > @sigmaaa
> > Tested under: OS - Linux ESP-IDF: master
> > I was not even able to list hints. (Window -> Show View -> Other... -> Hints) Please, check.
> > ![image](https://user-images.githubusercontent.com/69584713/186365231-967179fc-b4b2-473d-b040-e1cf9f636340.png)
> 
> Hi @AndriiFilippov, I pushed a new commit. Should work on Linux now. However, I noticed, that the hints.yml structure is updated on the latest esp-idf master. I need to make some changes to support the new hints.yml file. I'll keep you updated. Thanks

Hi @sigmaaa, tested on Linux again. Works ! 👍

kolipakakondal commented:
Hi @sigmaaa I'm on esp-idf v.4.4 

What is the default behavior in this case? I know hints.yml is supported only from the v5.0 but would you like to show a message on the viewer  - `Hints are only supported from esp-idf v5.0 and higher` or something like this? Otherwise, the user might be under the impression that it's not working.

<img width="1148" alt="Screenshot 2022-09-01 at 10 23 58 AM" src="https://user-images.githubusercontent.com/8463287/187834226-4a1c0a1b-290e-43b2-8698-3b95c2302e52.png">

kolipakakondal commented:
The filter does not seem to work based on the text provided. please find the screenshot

Also, we could filter the table entries as the user start type-in if the exact error message is not found

<img width="650" alt="Screenshot 2022-09-01 at 10 38 55 AM" src="https://user-images.githubusercontent.com/8463287/187835889-e97dd69b-f30e-4a63-8527-23b50e557f08.png">

kolipakakondal commented:
Could you please check the below error messages reported in the error log?

<img width="458" alt="Screenshot 2022-09-01 at 10 42 09 AM" src="https://user-images.githubusercontent.com/8463287/187836258-c0f570df-1832-4b8e-9adb-9b1c1215b178.png">

`
java.util.regex.PatternSyntaxException: Illegal repetition near index 112
format '([^']+)' expects argument of type '((unsigned )?int|long)', but argument (\w+) has type '([u]?int32_t)'( {aka '([^']+)'})?
                                                                                                                ^
	at java.base/java.util.regex.Pattern.error(Pattern.java:2027)
	at java.base/java.util.regex.Pattern.closure(Pattern.java:3320)
	at java.base/java.util.regex.Pattern.sequence(Pattern.java:2213)
	at java.base/java.util.regex.Pattern.expr(Pattern.java:2068)
	at java.base/java.util.regex.Pattern.group0(Pattern.java:3057)
	at java.base/java.util.regex.Pattern.sequence(Pattern.java:2123)
	at java.base/java.util.regex.Pattern.expr(Pattern.java:2068)
	at java.base/java.util.regex.Pattern.compile(Pattern.java:1782)
	at java.base/java.util.regex.Pattern.<init>(Pattern.java:1428)
	at java.base/java.util.regex.Pattern.compile(Pattern.java:1068)
	at com.espressif.idf.ui.dialogs.HintsView$4.modifyText(HintsView.java:156)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:183)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1522)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1545)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1530)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1324)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4310)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3883)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1041)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
`

kolipakakondal commented on bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/HintsUtil.java:
```diff
@@ -0,0 +1,54 @@
+
+		File hintsYmFile = new File(IDFUtil.getIDFPath() + File.separator + "tools" + File.separator + "idf_py_actions"
+				+ File.separator
+				+ "hints.yml"); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
```
texts and NON-NLS seems to be on the different lines

sigmaaa replied:
fixed

kolipakakondal commented on bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/HintsUtil.java:
```diff
@@ -0,0 +1,54 @@
+		List<String[]> reHintsPairArray = new ArrayList<>();
+		try
+		{
+			reHintsPairArray = hintsYmFile.isFile() ? parseHintsYamlFile(hintsYmFile) : reHintsPairArray;
```
It would be great if you could log whether hints.yml file is found or not and the path for it for troubleshooting purposes

sigmaaa replied:
done

kolipakakondal commented on bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/HintsUtil.java:
```diff
@@ -0,0 +1,54 @@
+		List<String[]> reHintsPairArray = new ArrayList<>();
+		InputStream inputStream;
+		inputStream = new FileInputStream(hintsYmFile);
+		ArrayList<LinkedHashMap<String, String>> reHintsArray = yaml.load(inputStream);
```
We could replace with super type `List`

sigmaaa replied:
fixed

kolipakakondal commented on bundles/com.espressif.idf.ui/plugin.xml:
```diff
@@ -565,6 +565,13 @@
+            id="com.espressif.idf.ui.views.hints"
+            name="%hints_view.name"
+            restorable="true">
+      </view>
```
Could you create an Espressif category for views and add Hints as part of it. I think Ali added Callers View previously, we could put these two items as part of the Espressif view. 

This would help users to find it easily

sigmaaa replied:
done

sigmaaa commented:
> Hi @sigmaaa I'm on esp-idf v.4.4
> 
> What is the default behavior in this case? I know hints.yml is supported only from the v5.0 but would you like to show a message on the viewer - `Hints are only supported from esp-idf v5.0 and higher` or something like this? Otherwise, the user might be under the impression that it's not working.
> 
> <img alt="Screenshot 2022-09-01 at 10 23 58 AM" width="1148" src="https://user-images.githubusercontent.com/8463287/187834226-4a1c0a1b-290e-43b2-8698-3b95c2302e52.png">

Hi Kondal, thanks for the review. I pushed the new commit with providing the error msg to the hints view

sigmaaa commented:
> Could you please check the below error messages reported in the error log?
> 
> <img alt="Screenshot 2022-09-01 at 10 42 09 AM" width="458" src="https://user-images.githubusercontent.com/8463287/187836258-c0f570df-1832-4b8e-9adb-9b1c1215b178.png">
> 
> `java.util.regex.PatternSyntaxException: Illegal repetition near index 112 format '([^']+)' expects argument of type '((unsigned )?int|long)', but argument (\w+) has type '([u]?int32_t)'( {aka '([^']+)'})? ^ at java.base/java.util.regex.Pattern.error(Pattern.java:2027) at java.base/java.util.regex.Pattern.closure(Pattern.java:3320) at java.base/java.util.regex.Pattern.sequence(Pattern.java:2213) at java.base/java.util.regex.Pattern.expr(Pattern.java:2068) at java.base/java.util.regex.Pattern.group0(Pattern.java:3057) at java.base/java.util.regex.Pattern.sequence(Pattern.java:2123) at java.base/java.util.regex.Pattern.expr(Pattern.java:2068) at java.base/java.util.regex.Pattern.compile(Pattern.java:1782) at java.base/java.util.regex.Pattern.<init>(Pattern.java:1428) at java.base/java.util.regex.Pattern.compile(Pattern.java:1068) at com.espressif.idf.ui.dialogs.HintsView$4.modifyText(HintsView.java:156) at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:183) at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89) at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4524) at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1522) at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1545) at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1530) at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1324) at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4310) at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3883) at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150) at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338) at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1041) at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155) at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644) at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338) at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551) at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156) at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152) at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203) at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136) at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104) at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401) at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.base/java.lang.reflect.Method.invoke(Method.java:566) at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659) at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596) at org.eclipse.equinox.launcher.Main.run(Main.java:1467) at org.eclipse.equinox.launcher.Main.main(Main.java:1440)`



> The filter does not seem to work based on the text provided. please find the screenshot
> 
> Also, we could filter the table entries as the user start type-in if the exact error message is not found
> 
> <img alt="Screenshot 2022-09-01 at 10 38 55 AM" width="650" src="https://user-images.githubusercontent.com/8463287/187835889-e97dd69b-f30e-4a63-8527-23b50e557f08.png">

In this case, the filter does not work because of the `Unhandled event loop exception`, which happened in the compilation regular expression stage. But filtering the table entries as the user start type-in if the exact error message is not found is a good idea. I added this functionality in the latest commit. Thanks

sigmaaa commented:
Hi @AndriiFilippov, @kolipakakondal. 

I added a few commits with refactoring and improving some cases. Please, take a look, when you have time. Summary:
- Support for the new structure of the hints.yml, which contains `re_variables` and `hints_variables`
- Added the error message to the hints viewer, when hints.yml is missing
- Added information icon to the hints column for visual separation of the table columns
- Filtering the table entries as the user starts type-in if the exact error message is not found
- Added Espressif category for Callers and Hints views 

Unhandled event loop exception happens because of the mistake in the hints.yml file. It has to be fixed on the esp-idf side. 
For now, please open the `hints.yml` file, which is located in 'esp-idf/tools/idf_py_actions' directory, and replace this line:
`re: "format '([^']+)' expects argument of type '((unsigned )?int|long)', but argument (\\w+) has type '([u]?int32_t)'( {aka '([^']+)'})?"`
with this
`re: "format '([^']+)' expects argument of type '((unsigned )?int|long)', but argument (\\w+) has type '([u]?int32_t)'( \\{aka '([^']+)'\\})?"`
